### PR TITLE
ipt: implement IPT functions in libdrakvuf

### PR DESCRIPTION
**Pull request by icedevml**
Provide API in libdrakvuf such that plugins could interact with Intel Processor Trace feature.

### CODE REVIEW

**Comment by icedevml** [comment]
ready

File: src/libdrakvuf/drakvuf.c [hunk 1]
```diff
@@ -879,3 +879,37 @@ bool drakvuf_set_vcpu_gprs(drakvuf_t drakvuf, int vcpu, registers_t* regs)
...
+        *size = drakvuf->ipt_state[vcpu].size;
+        return true;
+    }
+
+    return false;
+}
+
+bool drakvuf_get_ipt_offset(drakvuf_t drakvuf, int vcpu, uint64_t* offset, uint64_t* last_offset)
+{
+    int rc = xen_get_ipt_offset(drakvuf->xen, drakvuf->domID, vcpu, &drakvuf->ipt_state[vcpu]);
```
Comment (tklengyel) [inline src/libdrakvuf/drakvuf.c:899]: I don't think we need `int rc` here, just do an `if ( ... ) return false`

File: src/libdrakvuf/drakvuf.c [hunk 2]
```diff
@@ -879,3 +879,37 @@ bool drakvuf_set_vcpu_gprs(drakvuf_t drakvuf, int vcpu, registers_t* regs)
...
+        *last_offset = drakvuf->ipt_state[vcpu].last_offset;
+        return true;
+    }
+
+    return false;
+}
+
+bool drakvuf_disable_ipt(drakvuf_t drakvuf, int vcpu)
+{
+    int rc = xen_disable_ipt(drakvuf->xen, drakvuf->domID, vcpu, &drakvuf->ipt_state[vcpu]);
```
Comment (tklengyel) [inline src/libdrakvuf/drakvuf.c:913]: Same, no need for `int rc`, you can just return `xen_disable_ipt` right away, it's already effectively a bool.

File: src/libdrakvuf/drakvuf.c [hunk 3]
```diff
@@ -879,3 +879,37 @@ bool drakvuf_set_vcpu_gprs(drakvuf_t drakvuf, int vcpu, registers_t* regs)
 
     return xen_set_vcpu_ctx(drakvuf->xen, drakvuf->domID, vcpu, &ctx);
 }
+
+bool drakvuf_enable_ipt(drakvuf_t drakvuf, int vcpu, uint8_t** buf, uint64_t* size)
+{
+    int rc = xen_enable_ipt(drakvuf->xen, drakvuf->domID, vcpu, &drakvuf->ipt_state[vcpu]);
```
Comment (tklengyel) [inline src/libdrakvuf/drakvuf.c:885]: Same here, `if ( !xen_enable_ipt(...) ) return false;`

File: src/libdrakvuf/private.h [hunk 4]
```diff
@@ -245,6 +245,8 @@ struct drakvuf
     int event_fd_cnt;          // auto-generated for poll()
     fd_info_t fd_info_lookup;  // auto-generated for fast drakvuf_loop lookups
     int poll_rc;
+
+    ipt_state_t ipt_state[16]; // TODO dynamically allocate
```
Comment (icedevml) [inline src/libdrakvuf/private.h:249]: :thinking: I've forgot about this, but anyway there is a cap on 16 vCPUs somewhere else:

https://github.com/tklengyel/drakvuf/blob/master/src/libdrakvuf/vmi.c#L1516-L1517

are we OK with leaving static array for now?

File: src/libdrakvuf/drakvuf.c (hunk 1 above)
Comment (icedevml) [reply src/libdrakvuf/drakvuf.c]: fixed

File: src/libdrakvuf/private.h (hunk 4 above)
Comment (tklengyel) [reply src/libdrakvuf/private.h:249]: Yea, we have a 16-vcpu limit right now

Comment (icedevml) [reply src/libdrakvuf/private.h]: I've removed this `//TODO` for now and added bound check whether `vcpu < 16`

File: src/libdrakvuf/drakvuf.c [hunk 5]
```diff
@@ -879,3 +879,47 @@ bool drakvuf_set_vcpu_gprs(drakvuf_t drakvuf, int vcpu, registers_t* regs)
 
     return xen_set_vcpu_ctx(drakvuf->xen, drakvuf->domID, vcpu, &ctx);
 }
+
+static bool is_valid_vcpu(drakvuf_t drakvuf, int vcpu)
```
Comment (tklengyel) [inline src/libdrakvuf/drakvuf.c:883]: Any reason why the function takes `vcpu` as an `int`?

Comment (icedevml) [reply src/libdrakvuf/drakvuf.c:883]: Same as
```
bool drakvuf_set_vcpu_gprs(drakvuf_t drakvuf, int vcpu, registers_t* regs)
```

Comment (tklengyel) [reply src/libdrakvuf/drakvuf.c:883]: Yea, I think that should also be an unsigned int :)

Comment (icedevml) [reply src/libdrakvuf/drakvuf.c:883]: ok I will fix this

**Comment by icedevml** [comment]
@drakvuf-jenkins Retest this please.

File: src/libdrakvuf/drakvuf.c (hunk 5 above)
Comment (icedevml) [reply src/libdrakvuf/drakvuf.c]: fixed, the rest is fixed with #1007

File: src/libdrakvuf/drakvuf.c [hunk 6]
```diff
@@ -879,3 +879,49 @@ bool drakvuf_set_vcpu_gprs(drakvuf_t drakvuf, int vcpu, registers_t* regs)
 
     return xen_set_vcpu_ctx(drakvuf->xen, drakvuf->domID, vcpu, &ctx);
 }
+
+static bool is_valid_vcpu(drakvuf_t drakvuf, unsigned int vcpu)
+{
+    // VMs with more than 16 vCPUs are not supported for usage with IPT
+    // this limit is DRAKVUF specific
+    return vcpu < 16 && drakvuf->vcpus > vcpu;
```
Comment (tklengyel) [inline src/libdrakvuf/drakvuf.c:887]: So I think we should declare `16` as a #define in `private.h` as it is now being used at more spots, say `#define MAX_DRAKVUF_VCPU 16`. That way we can choose to increase it any time fairly easily in the future. It is a random number at the moment, there is nothing technically limiting us to 16, it's just I think a high enough number that will be unlikely to get exceeded for any reasonable usecase.

Comment (icedevml) [reply src/libdrakvuf/drakvuf.c:887]: ok

Comment (icedevml) [reply src/libdrakvuf/drakvuf.c]: done

**Comment by icedevml** [comment]
@drakvuf-jenkins Retest this please.

**Comment by icedevml** [comment]
@drakvuf-jenkins Retest this please.

**Pull request merged**